mono/io-layer/processes.c

```diff
@@ -2754,7 +2754,32 @@ process_wait (gpointer handle, guint32 timeout, gboolean alertable)
 	/* We don't need to lock mono_processes here, the entry
 	 * has a handle_count > 0 which means it will not be freed. */
 	mp = process_handle->mono_process;
-	g_assert (mp);
+	if (!mp) {
+		guint32 start, now;
+
+		/* We've already done this one */
+		MONO_TRACE (G_LOG_LEVEL_DEBUG, MONO_TRACE_IO_LAYER, "%s (%p, %u): waiting on non-child process", __func__, handle, timeout);
+
+		for (start = now = mono_msec_ticks (); timeout == INFINITE || now - start < timeout || now == start; now = mono_msec_ticks ()) {
+			int status;
+			pid_t res = waitpid (pid, &status, WNOHANG);
```

Comment:
This doesn't looks particularly related to the change, it better go to a separate commit. 

I'm not sure this is a behavior we want to support, I rather throw an exception and force the user code
to do the polling.